src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+            Ok(Edn::Rational(n))
+        }
+        _ => Err(Error::ParseEdn(format!("{} could not be parsed", number))),
+    }
+}
+
+fn read_char(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let c = chars.next();
+    c.ok_or(format!("{:?} could not be parsed", c))
+        .map(|c| Edn::Char(c))
+        .map_err(|e| Error::ParseEdn(e))
+}
+
+fn read_bool_or_nil(c: char, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    match c {
+        't' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'r' || e == &'u' || e == &'e')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let t = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&t);
+            Ok(Edn::Bool(string.parse::<bool>()?))
+        }
+        'f' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'a' || e == &'l' || e == &'s' || e == &'e')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let f = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&f);
+            Ok(Edn::Bool(string.parse::<bool>()?))
+        }
+        'n' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'i' || e == &'l')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let n = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&n);
+            match &string[..] {
+                "nil" => Ok(Edn::Nil),
+                _ => Err(Error::ParseEdn(format!("{} cound not be parsed", string))),
+            }
+        }
+        _ => Err(Error::ParseEdn(
+            "Nullable boolean cound not be parsed".to_string(),
+        )),
+    }
+}
+
+fn read_vec(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(']') => return Ok(Edn::Vector(Vector::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_list(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(')') => return Ok(Edn::List(List::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
```

Comment:
Again, I think this second `Some(c)` is not doing anything, it could be removed.

Reply:
failures:
    deserialize::parse::test::parse_complex
    deserialize::parse::test::parse_list
    deserialize::test::from_str_list_with_set
    deserialize::test::from_str_list_with_vec

Reply:
I actually meant that the `err` should have this logic, but it's the same as this.